<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="singlecellexperiment" href="api/modules.html" /><link rel="prev" title="SingleCellExperiment" href="index.html" />

    <!-- Generated with Sphinx 8.1.3 and Furo 2024.08.06 -->
        <title>Represent single-cell experiments - SingleCellExperiment 0.5.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #eeffcc;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">SingleCellExperiment 0.5.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon no-toc" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">SingleCellExperiment 0.5.1 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Represent single-cell experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="#construction">Construction</a></li>
<li class="toctree-l1"><a class="reference internal" href="#getters-setters">Getters/Setters</a></li>
<li class="toctree-l1"><a class="reference internal" href="#subset-experiments">Subset experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="#combining-experiments">Combining experiments</a></li>
<li class="toctree-l1"><a class="reference internal" href="#export-as-anndata-or-mudata">Export as <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> or <code class="docutils literal notranslate"><span class="pre">MuData</span></code></a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="api/modules.html">Module Reference</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of Module Reference</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l2 has-children"><a class="reference internal" href="api/singlecellexperiment.html">singlecellexperiment package</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" role="switch" type="checkbox"/><label for="toctree-checkbox-2"><div class="visually-hidden">Toggle navigation of singlecellexperiment package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul>
<li class="toctree-l3"><a class="reference internal" href="api/singlecellexperiment.io.html">singlecellexperiment.io package</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/tutorial.md.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon no-toc" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="represent-single-cell-experiments">
<h1>Represent single-cell experiments<a class="headerlink" href="#represent-single-cell-experiments" title="Link to this heading">¶</a></h1>
<p>This package provides container class to represent single-cell experimental data as 2-dimensional matrices. In these matrices, the rows typically denote features or genomic regions of interest, while columns represent cells. In addition, a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> (SCE) object may contain low-dimensionality embeddings, alternative experiments performed on same sample or set of cells.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>The design of <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> class and its derivates adheres to the R/Bioconductor specification, where rows correspond to features, and columns represent cells.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These classes follow a functional paradigm for accessing or setting properties, with further details discussed in <a class="reference external" href="https://biocpy.github.io/tutorial/chapters/philosophy.html#functional-discipline">functional paradigm</a> section.</p>
</div>
</section>
<section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Link to this heading">¶</a></h1>
<p>To get started, install the package from <a class="reference external" href="https://pypi.org/project/singlecellexperiment/">PyPI</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip<span class="w"> </span>install<span class="w"> </span>singlecellexperiment
</pre></div>
</div>
</section>
<section id="construction">
<h1>Construction<a class="headerlink" href="#construction" title="Link to this heading">¶</a></h1>
<p>The <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> extends <code class="docutils literal notranslate"><span class="pre">RangeSummarizedExperiment</span></code> and contains additional attributes:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">reduced_dims</span></code>: Slot for low-dimensionality embeddings for each cell.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">alternative_experiments</span></code>: Manages multi-modal experiments performed on the same sample or set of cells.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">row_pairs</span></code> or <code class="docutils literal notranslate"><span class="pre">column_pairs</span></code>: Stores relationships between features or cells.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In contrast to R, matrices in Python are unnamed and do not contain row or column names. Hence, these matrices cannot be directly used as values in assays or alternative experiments. We strictly enforce type checks in these cases. To relax these restrictions for alternative experiments, set <code class="docutils literal notranslate"><span class="pre">type_check_alternative_experiments</span></code> to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
</div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>If you are using the <code class="docutils literal notranslate"><span class="pre">alternative_experiments</span></code> slot, the number of cells must match the parent experiment.  Otherwise, the expectation is that the cells do not share the same sample or annotations and cannot be set in alternative experiments!</p>
</div>
<p>Before we construct a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> object, lets generate information about rows, columns and a mock experimental data from single-cell rna-seq experiments:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">sparse</span> <span class="k">as</span> <span class="n">sp</span>
<span class="kn">from</span> <span class="nn">biocframe</span> <span class="kn">import</span> <span class="n">BiocFrame</span>
<span class="kn">from</span> <span class="nn">genomicranges</span> <span class="kn">import</span> <span class="n">GenomicRanges</span>
<span class="kn">from</span> <span class="nn">random</span> <span class="kn">import</span> <span class="n">random</span>

<span class="n">nrows</span> <span class="o">=</span> <span class="mi">200</span>
<span class="n">ncols</span> <span class="o">=</span> <span class="mi">6</span>
<span class="n">counts</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="p">,</span> <span class="n">density</span><span class="o">=</span><span class="mf">0.2</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;csr&quot;</span><span class="p">)</span>
<span class="n">row_data</span> <span class="o">=</span> <span class="n">BiocFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr1&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chr3&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="o">*</span> <span class="mi">20</span><span class="p">,</span>
        <span class="s2">&quot;starts&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">300</span><span class="p">),</span>
        <span class="s2">&quot;ends&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">110</span><span class="p">,</span> <span class="mi">310</span><span class="p">),</span>
        <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;+&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">20</span><span class="p">,</span>
        <span class="s2">&quot;score&quot;</span><span class="p">:</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">200</span><span class="p">),</span>
        <span class="s2">&quot;GC&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">random</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">)]</span> <span class="o">*</span> <span class="mi">20</span><span class="p">,</span>
    <span class="p">}</span>
<span class="p">)</span>

<span class="n">col_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;celltype&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;cluster1&quot;</span><span class="p">,</span> <span class="s2">&quot;cluster2&quot;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now lets create the <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> instance:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">singlecellexperiment</span> <span class="kn">import</span> <span class="n">SingleCellExperiment</span>

<span class="n">sce</span> <span class="o">=</span> <span class="n">SingleCellExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">counts</span><span class="p">},</span> <span class="n">row_data</span><span class="o">=</span><span class="n">row_data</span><span class="p">,</span> <span class="n">column_data</span><span class="o">=</span><span class="n">col_data</span><span class="p">,</span>
    <span class="n">reduced_dims</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;random_embeds&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">(</span><span class="n">ncols</span><span class="p">,</span> <span class="mi">4</span><span class="p">)}</span>
<span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment
dimensions: (200, 6)
assays(1): [&#39;counts&#39;]
row_data columns(6): [&#39;seqnames&#39;, &#39;starts&#39;, &#39;ends&#39;, &#39;strand&#39;, &#39;score&#39;, &#39;GC&#39;]
row_names(0):  
column_data columns(1): [&#39;celltype&#39;]
column_names(6): [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, &#39;3&#39;, &#39;4&#39;, &#39;5&#39;]
main_experiment_name:  
reduced_dims(1): [&#39;random_embeds&#39;]
alternative_experiments(0): []
row_pairs(0): []
column_pairs(0): []
metadata(0): 
</pre></div>
</div>
</div>
</div>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>You can also use delayed or file-backed arrays for representing experimental data, check out <a class="reference internal" href="#./summarized_experiment.qmd#delayed-or-file-backed-arrays"><span class="xref myst">this section</span></a> from summarized experiment.</p>
</div>
<section id="interop-with-anndata">
<h2>Interop with <code class="docutils literal notranslate"><span class="pre">anndata</span></code><a class="headerlink" href="#interop-with-anndata" title="Link to this heading">¶</a></h2>
<p>We provide convenient methods for loading an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> or <code class="docutils literal notranslate"><span class="pre">h5ad</span></code> file into <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects.</p>
<p>For example, lets create an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">sparse</span> <span class="k">as</span> <span class="n">sp</span>

<span class="n">counts</span> <span class="o">=</span> <span class="n">sp</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 100 × 2000
</pre></div>
</div>
</div>
</div>
<p>Converting <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> as <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> is straightforward:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sce_adata</span> <span class="o">=</span> <span class="n">SingleCellExperiment</span><span class="o">.</span><span class="n">from_anndata</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce_adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment
dimensions: (2000, 100)
assays(1): [&#39;X&#39;]
row_data columns(0): []
row_names(2000): [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, ..., &#39;1997&#39;, &#39;1998&#39;, &#39;1999&#39;]
column_data columns(0): []
column_names(100): [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;, ..., &#39;97&#39;, &#39;98&#39;, &#39;99&#39;]
main_experiment_name:  
reduced_dims(0): []
alternative_experiments(0): []
row_pairs(0): []
column_pairs(0): []
metadata(1): uns
</pre></div>
</div>
</div>
</div>
<p>and vice-verse.  All assays from SCE are represented in the <code class="docutils literal notranslate"><span class="pre">layers</span></code> slot of the <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata2</span> <span class="o">=</span> <span class="n">sce_adata</span><span class="o">.</span><span class="n">to_anndata</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">adata2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(AnnData object with n_obs × n_vars = 100 × 2000
    uns: &#39;uns&#39;
    layers: &#39;X&#39;, None)
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
</div>
</div>
<p>Similarly, one can load a h5ad file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">singlecellexperiment</span> <span class="kn">import</span> <span class="n">read_h5ad</span>
<span class="n">sce_h5</span> <span class="o">=</span> <span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;../../assets/data/adata.h5ad&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce_h5</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="from-10x-formats">
<h2>From 10X formats<a class="headerlink" href="#from-10x-formats" title="Link to this heading">¶</a></h2>
<p>In addition, we also provide convenient methods to load a <a class="reference external" href="https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-h5-matrices">10X Genomics HDF5 Feature-Barcode Matrix Format</a> file.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">singlecellexperiment</span> <span class="kn">import</span> <span class="n">read_tenx_h5</span>
<span class="n">sce_h5</span> <span class="o">=</span> <span class="n">read_tenx_h5</span><span class="p">(</span><span class="s2">&quot;../../assets/data/tenx.sub.h5&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce_h5</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Methods are also available to read a 10x matrix market directory using the <code class="docutils literal notranslate"><span class="pre">read_tenx_mtx</span></code> function.</p>
</div>
</section>
</section>
<section id="getters-setters">
<h1>Getters/Setters<a class="headerlink" href="#getters-setters" title="Link to this heading">¶</a></h1>
<p>Getters are available to access various attributes using either the property notation or functional style.</p>
<p>Since <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> extends <code class="docutils literal notranslate"><span class="pre">RangedSummarizedExperiment</span></code>, all getters and setters from the base class are accessible here; more details <a class="reference internal" href="#./summarized_experiment.qmd"><span class="xref myst">here</span></a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># access assay names</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reduced dim names (as property): &quot;</span><span class="p">,</span> <span class="n">sce</span><span class="o">.</span><span class="n">reduced_dim_names</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;reduced dim names (functional style): &quot;</span><span class="p">,</span> <span class="n">sce</span><span class="o">.</span><span class="n">get_reduced_dim_names</span><span class="p">())</span>

<span class="c1"># access row data</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce</span><span class="o">.</span><span class="n">row_data</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>reduced dim names (as property):  [&#39;random_embeds&#39;]
reduced dim names (functional style):  [&#39;random_embeds&#39;]
BiocFrame with 200 rows and 6 columns
      seqnames  starts    ends strand   score                  GC
        &lt;list&gt; &lt;range&gt; &lt;range&gt; &lt;list&gt; &lt;range&gt;              &lt;list&gt;
  [0]     chr1     100     110      -       0  0.4846466442263945
  [1]     chr2     101     111      +       1  0.6393252758758694
  [2]     chr2     102     112      +       2 0.08113519128304914
           ...     ...     ...    ...     ...                 ...
[197]     chr3     297     307      +     197  0.6773244982271057
[198]     chr3     298     308      -     198  0.9504434474906067
[199]     chr3     299     309      -     199  0.9375884174889054
</pre></div>
</div>
</div>
</div>
<section id="access-a-reduced-dimension">
<h2>Access a reduced dimension<a class="headerlink" href="#access-a-reduced-dimension" title="Link to this heading">¶</a></h2>
<p>One can access an reduced dimension by index or name:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sce</span><span class="o">.</span><span class="n">reduced_dim</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span> <span class="c1"># same as se.reduced_dim(&quot;random_embeds&quot;)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[0.71689997, 0.20085595, 0.10838405, 0.92681804],
       [0.21441565, 0.11274687, 0.86120723, 0.28558661],
       [0.06031193, 0.42221772, 0.93394306, 0.70542632],
       [0.17739584, 0.11835883, 0.76782622, 0.6049811 ],
       [0.45533928, 0.51673548, 0.0290028 , 0.75461485],
       [0.68960122, 0.72903642, 0.79967072, 0.06252843]])
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="subset-experiments">
<h1>Subset experiments<a class="headerlink" href="#subset-experiments" title="Link to this heading">¶</a></h1>
<p>You can subset experimental data by using the subset (<code class="docutils literal notranslate"><span class="pre">[]</span></code>) operator. This operation accepts different slice input types, such as a boolean vector, a <code class="docutils literal notranslate"><span class="pre">slice</span></code> object, a list of indices, or names (if available) to subset.</p>
<p>In our previous example, we didn’t include row or column names. Let’s create another <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> object that includes names.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">subset_sce</span> <span class="o">=</span> <span class="n">sce</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">subset_sce</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment
dimensions: (10, 3)
assays(1): [&#39;counts&#39;]
row_data columns(6): [&#39;seqnames&#39;, &#39;starts&#39;, &#39;ends&#39;, &#39;strand&#39;, &#39;score&#39;, &#39;GC&#39;]
row_names(0):  
column_data columns(1): [&#39;celltype&#39;]
column_names(3): [&#39;0&#39;, &#39;1&#39;, &#39;2&#39;]
main_experiment_name:  
reduced_dims(1): [&#39;random_embeds&#39;]
alternative_experiments(0): []
row_pairs(0): []
column_pairs(0): []
metadata(0): 
</pre></div>
</div>
</div>
</div>
</section>
<section id="combining-experiments">
<h1>Combining experiments<a class="headerlink" href="#combining-experiments" title="Link to this heading">¶</a></h1>
<p><code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> implements methods for the <code class="docutils literal notranslate"><span class="pre">combine</span></code> generic from <a class="reference external" href="https://github.com/BiocPy/biocutils"><strong>BiocUtils</strong></a>.</p>
<p>These methods enable the merging or combining of multiple <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects, allowing users to aggregate data from different experiments or conditions. Note: <code class="docutils literal notranslate"><span class="pre">row_pairs</span></code> and <code class="docutils literal notranslate"><span class="pre">column_pairs</span></code> are not ignored as part of this operation.</p>
<p>To demonstrate, let’s create multiple <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> objects (read more about this in <a class="reference internal" href="#./summarized_experiment.qmd#combining-experiments"><span class="xref myst">combine section from <code class="docutils literal notranslate"><span class="pre">SummarizedExperiment</span></code></span></a>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ncols</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">nrows</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">sce1</span> <span class="o">=</span> <span class="n">SingleCellExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="p">))},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">BiocFrame</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">nrows</span><span class="p">}),</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">BiocFrame</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">ncols</span><span class="p">}),</span>
<span class="p">)</span>

<span class="n">sce2</span> <span class="o">=</span> <span class="n">SingleCellExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">nrows</span><span class="p">,</span> <span class="n">ncols</span><span class="p">)),</span>
        <span class="c1"># &quot;normalized&quot;: np.random.normal(size=(nrows, ncols)),</span>
    <span class="p">},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">BiocFrame</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">*</span> <span class="n">nrows</span><span class="p">}),</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">BiocFrame</span><span class="p">({</span><span class="s2">&quot;A&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">*</span> <span class="n">ncols</span><span class="p">}),</span>
<span class="p">)</span>

<span class="n">rowdata1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr_5&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_3&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">500</span><span class="p">,</span> <span class="mi">300</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span>
        <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">510</span><span class="p">,</span> <span class="mi">310</span><span class="p">,</span> <span class="mi">210</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HER2&quot;</span><span class="p">,</span> <span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;TPFK&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">coldata1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SAM_1&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_2&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_3&quot;</span><span class="p">],</span>
        <span class="s2">&quot;disease&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">],</span>
        <span class="s2">&quot;doublet_score&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.15</span><span class="p">,</span> <span class="mf">0.62</span><span class="p">,</span> <span class="mf">0.18</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_1&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_2&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_3&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">sce_alts1</span> <span class="o">=</span> <span class="n">SingleCellExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
        <span class="s2">&quot;lognorm&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">lognormal</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
    <span class="p">},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">rowdata1</span><span class="p">,</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">coldata1</span><span class="p">,</span>
    <span class="n">row_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HER2&quot;</span><span class="p">,</span> <span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;TPFK&quot;</span><span class="p">],</span>
    <span class="n">column_names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_1&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_2&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_3&quot;</span><span class="p">],</span>
    <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq_type&quot;</span><span class="p">:</span> <span class="s2">&quot;paired&quot;</span><span class="p">},</span>
    <span class="n">reduced_dims</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;PCA&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">))},</span>
    <span class="n">alternative_experiments</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;modality1&quot;</span><span class="p">:</span> <span class="n">SingleCellExperiment</span><span class="p">(</span>
            <span class="n">assays</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts2&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))},</span>
        <span class="p">)</span>
    <span class="p">},</span>
<span class="p">)</span>

<span class="n">rowdata2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;seqnames&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;chr_5&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_3&quot;</span><span class="p">,</span> <span class="s2">&quot;chr_2&quot;</span><span class="p">],</span>
        <span class="s2">&quot;start&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">500</span><span class="p">,</span> <span class="mi">300</span><span class="p">,</span> <span class="mi">200</span><span class="p">],</span>
        <span class="s2">&quot;end&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">510</span><span class="p">,</span> <span class="mi">310</span><span class="p">,</span> <span class="mi">210</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;HER2&quot;</span><span class="p">,</span> <span class="s2">&quot;BRCA1&quot;</span><span class="p">,</span> <span class="s2">&quot;TPFK&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">coldata2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="p">{</span>
        <span class="s2">&quot;sample&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;SAM_4&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_5&quot;</span><span class="p">,</span> <span class="s2">&quot;SAM_6&quot;</span><span class="p">],</span>
        <span class="s2">&quot;disease&quot;</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;True&quot;</span><span class="p">,</span> <span class="s2">&quot;False&quot;</span><span class="p">,</span> <span class="s2">&quot;True&quot;</span><span class="p">],</span>
        <span class="s2">&quot;doublet_score&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.23</span><span class="p">,</span> <span class="mf">0.54</span><span class="p">],</span>
    <span class="p">},</span>
    <span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_4&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_5&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_6&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">sce_alts2</span> <span class="o">=</span> <span class="n">SingleCellExperiment</span><span class="p">(</span>
    <span class="n">assays</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;counts&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">)),</span>
        <span class="c1"># &quot;lognorm&quot;: np.random.lognormal(size=(3, 3)),</span>
    <span class="p">},</span>
    <span class="n">row_data</span><span class="o">=</span><span class="n">rowdata2</span><span class="p">,</span>
    <span class="n">column_data</span><span class="o">=</span><span class="n">coldata2</span><span class="p">,</span>
    <span class="n">metadata</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;seq_platform&quot;</span><span class="p">:</span> <span class="s2">&quot;Illumina NovaSeq 6000&quot;</span><span class="p">},</span>
    <span class="n">reduced_dims</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;PCA&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">))},</span>
    <span class="n">alternative_experiments</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;modality1&quot;</span><span class="p">:</span> <span class="n">SingleCellExperiment</span><span class="p">(</span>
            <span class="n">assays</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;counts2&quot;</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">poisson</span><span class="p">(</span><span class="n">lam</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))},</span>
        <span class="p">)</span>
    <span class="p">},</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">combine_rows</span></code> or <code class="docutils literal notranslate"><span class="pre">combine_columns</span></code> operations, expect all experiments to contain the same assay names. To combine experiments by row:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">biocutils</span> <span class="kn">import</span> <span class="n">relaxed_combine_columns</span><span class="p">,</span> <span class="n">combine_columns</span><span class="p">,</span> <span class="n">combine_rows</span><span class="p">,</span> <span class="n">relaxed_combine_rows</span>
<span class="n">sce_combined</span> <span class="o">=</span> <span class="n">combine_rows</span><span class="p">(</span><span class="n">sce2</span><span class="p">,</span> <span class="n">sce1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce_combined</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment
dimensions: (200, 10)
assays(1): [&#39;counts&#39;]
row_data columns(1): [&#39;A&#39;]
row_names(0):  
column_data columns(1): [&#39;A&#39;]
column_names(0):  
main_experiment_name:  
reduced_dims(0): []
alternative_experiments(0): []
row_pairs(0): []
column_pairs(0): []
metadata(0): 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/singlecellexperiment/SingleCellExperiment.py:1066: UserWarning: &#39;row_pairs&#39; and &#39;column_pairs&#39; are currently ignored during this operation.
  warn(
</pre></div>
</div>
</div>
</div>
<p>Similarly to combine by column:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sce_combined</span> <span class="o">=</span> <span class="n">combine_columns</span><span class="p">(</span><span class="n">sce2</span><span class="p">,</span> <span class="n">sce1</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce_combined</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment
dimensions: (100, 20)
assays(1): [&#39;counts&#39;]
row_data columns(1): [&#39;A&#39;]
row_names(0):  
column_data columns(1): [&#39;A&#39;]
column_names(0):  
main_experiment_name:  
reduced_dims(0): []
alternative_experiments(0): []
row_pairs(0): []
column_pairs(0): []
metadata(0): 
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/singlecellexperiment/SingleCellExperiment.py:1108: UserWarning: &#39;row_pairs&#39; and &#39;column_pairs&#39; are currently ignored during this operation.
  warn(
</pre></div>
</div>
</div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can use <code class="docutils literal notranslate"><span class="pre">relaxed_combine_columns</span></code> or <code class="docutils literal notranslate"><span class="pre">relaxed_combined_rows</span></code> when there’s mismatch in the number of features or samples. Missing rows or columns in any object are filled in with appropriate placeholder values before combining, e.g. missing assay’s are replaced with a masked numpy array.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># sce_alts1 contains an additional assay not present in sce_alts2</span>
<span class="n">sce_relaxed_combine</span> <span class="o">=</span> <span class="n">relaxed_combine_columns</span><span class="p">(</span><span class="n">sce_alts1</span><span class="p">,</span> <span class="n">sce_alts2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sce_relaxed_combine</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>class: SingleCellExperiment
dimensions: (3, 6)
assays(2): [&#39;lognorm&#39;, &#39;counts&#39;]
row_data columns(3): [&#39;seqnames&#39;, &#39;start&#39;, &#39;end&#39;]
row_names(3): [&#39;HER2&#39;, &#39;BRCA1&#39;, &#39;TPFK&#39;]
column_data columns(3): [&#39;sample&#39;, &#39;disease&#39;, &#39;doublet_score&#39;]
column_names(6): [&#39;cell_1&#39;, &#39;cell_2&#39;, &#39;cell_3&#39;, &#39;cell_4&#39;, &#39;cell_5&#39;, &#39;cell_6&#39;]
main_experiment_name:  
reduced_dims(1): [&#39;PCA&#39;]
alternative_experiments(1): [&#39;modality1&#39;]
row_pairs(0): []
column_pairs(0): []
metadata(1): seq_type
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/singlecellexperiment/SingleCellExperiment.py:1218: UserWarning: &#39;row_pairs&#39; and &#39;column_pairs&#39; are currently ignored during this operation.
  warn(&quot;&#39;row_pairs&#39; and &#39;column_pairs&#39; are currently ignored during this operation.&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="export-as-anndata-or-mudata">
<h1>Export as <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> or <code class="docutils literal notranslate"><span class="pre">MuData</span></code><a class="headerlink" href="#export-as-anndata-or-mudata" title="Link to this heading">¶</a></h1>
<p>The package also provides methods to convert a <code class="docutils literal notranslate"><span class="pre">SingleCellExperiment</span></code> object into a <code class="docutils literal notranslate"><span class="pre">MuData</span></code> representation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">mdata</span> <span class="o">=</span> <span class="n">sce</span><span class="o">.</span><span class="n">to_mudata</span><span class="p">()</span>
<span class="n">mdata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/mudata/_core/mudata.py:1531: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.
  self._update_attr(&quot;var&quot;, axis=0, join_common=join_common)
/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/mudata/_core/mudata.py:1429: FutureWarning: From 0.4 .update() will not pull obs/var columns from individual modalities by default anymore. Set mudata.set_options(pull_on_update=False) to adopt the new behaviour, which will become the default. Use new pull_obs/pull_var and push_obs/push_var methods for more flexibility.
  self._update_attr(&quot;obs&quot;, axis=1, join_common=join_common)
</pre></div>
</div>
<div class="output text_html"><pre>MuData object with n_obs × n_vars = 6 × 200
  var:	&#x27;seqnames&#x27;, &#x27;starts&#x27;, &#x27;ends&#x27;, &#x27;strand&#x27;, &#x27;score&#x27;, &#x27;GC&#x27;
  1 modality
    Unknown Modality:	6 x 200
      obs:	&#x27;celltype&#x27;, &#x27;rownames&#x27;
      var:	&#x27;seqnames&#x27;, &#x27;starts&#x27;, &#x27;ends&#x27;, &#x27;strand&#x27;, &#x27;score&#x27;, &#x27;GC&#x27;
      obsm:	&#x27;random_embeds&#x27;
      layers:	&#x27;counts&#x27;</pre></div></div>
</div>
<p>or coerce to an <code class="docutils literal notranslate"><span class="pre">AnnData</span></code> object:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="p">,</span> <span class="n">alts</span> <span class="o">=</span> <span class="n">sce</span><span class="o">.</span><span class="n">to_anndata</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;main experiment: &quot;</span><span class="p">,</span> <span class="n">adata</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;alternative experiments: &quot;</span><span class="p">,</span> <span class="n">alts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>main experiment:  AnnData object with n_obs × n_vars = 6 × 200
    obs: &#39;celltype&#39;, &#39;rownames&#39;
    var: &#39;seqnames&#39;, &#39;starts&#39;, &#39;ends&#39;, &#39;strand&#39;, &#39;score&#39;, &#39;GC&#39;
    obsm: &#39;random_embeds&#39;
    layers: &#39;counts&#39;
alternative experiments:  None
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>/home/runner/work/SingleCellExperiment/SingleCellExperiment/.tox/docs/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.
  warnings.warn(&quot;Transforming to str index.&quot;, ImplicitModificationWarning)
</pre></div>
</div>
</div>
</div>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="api/modules.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">singlecellexperiment</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="index.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Home</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, jkanche
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer no-toc">
      
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=b9afe91b"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>